AsyncCollectiveTensor: during tracing, logging should have no side effects

bdhirsh opened the pull request:
I noticed that when I added to print fw_metadata, the prints would fail. I think it had to do with the fact that printing an `AsyncCollectiveTensor(FakeTensor)` has some side effects.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118647
* #118646
* #118645
* #118644
* __->__ #118643
* #118937



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

bdhirsh committed:
AsyncCollectiveTensor: during tracing, logging should have no side effects

[ghstack-poisoned]

awgu commented on torch/distributed/_functional_collectives.py:
```diff
@@ -455,7 +456,11 @@ def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
+        if not is_fake(self.elem):
+            # If we have a real tensor, then we need to sync it before printing its data.
+            # but if we have a fake tensor, then we will likely be calling this method
+            # as tracing time while logging, so we don't want the logging to cause side effects.
```
maybe typo:
```suggestion
            # at tracing time while logging, so we don't want the logging to cause side effects.
```

wconstab commented on torch/distributed/_functional_collectives.py:
```diff
@@ -455,7 +456,11 @@ def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
 
     def __repr__(self):
-        wait_tensor(self.elem)
+        if not is_fake(self.elem):
```
hmm do we assume that if we are actually compiling, we will not be tracing into __repr__ bc a graph-break would have been triggered at some other layer?  otherwise, don't we need the graph-break to ensure __repr__ doesn't get baked into the compilation and instead gets called in between graphs at runtime?

bdhirsh replied:
Hmm I'm actually not sure what happens if dynamo sees this code. My guess would have been that dynamo already knows to graph break on `print()`, so it would never even see the `__repr__` call (unless someone calls it directly from user code which is hopefully rare?).

My intention with this change was:

(1) I have a [PR](https://github.com/pytorch/pytorch/pull/118646) that adds more logging AOTAutograd, and (I think) it was failing to log in some cases because there was an `AsyncCollectiveTensor(FakeTensor)` stashed in AOTAutograd's metadata, that it was unable to print. This PR was just trying to avoid that logging from causing any side effects (converting the AsyncCollectiveTensor(FakeTensor) into a FakeTensor during logging)

(2) In theory, if dynamo ends up inlining this function because it was called in user code, I think dynamo would probably see real tensors, and not fake tensors? So the behavior of this fn should be unchanged (we still wait()).

Let me know if that did/didn't answer your question!

wconstab replied:
I see. Well (1) makes sense I guess, but I'm less sure about (2)

bdhirsh committed:
Update on "AsyncCollectiveTensor: during tracing, logging should have no side effects"

I noticed that when I added to print fw_metadata, the prints would fail. I think it had to do with the fact that printing an `AsyncCollectiveTensor(FakeTensor)` has some side effects.




[ghstack-poisoned]
